engine/src/flutter/impeller/compiler/compiler_unittests.cc

```diff
@@ -123,6 +123,178 @@ TEST_P(CompilerTest, BindingBaseForFragShader) {
+    return FloatInfo(name, location, 1u, 2u);
+  }
+  static UniformInfo Vec3(const std::string& name, uint32_t location) {
+    return FloatInfo(name, location, 1u, 3u);
+  }
+  static UniformInfo Vec4(const std::string& name, uint32_t location) {
+    return FloatInfo(name, location, 1u, 4u);
+  }
+  static UniformInfo Mat4(const std::string& name, uint32_t location) {
+    return FloatInfo(name, location, 4u, 4u);
+  }
+
+  constexpr bool operator==(const UniformInfo& other) const {
+    return (uniform_name == other.uniform_name &&  //
+            location == other.location &&          //
+            type_name == other.type_name &&        //
+            columns == other.columns &&            //
+            vec_size == other.vec_size);
+  }
+
+ private:
+  static UniformInfo FloatInfo(const std::string& name,
+                               uint32_t location,
+                               uint32_t columns,
+                               uint32_t vec_size) {
+    return UniformInfo{
+        .uniform_name = name,
+        .location = location,
+        .type_name = "ShaderType::kFloat",
+        .columns = columns,
+        .vec_size = vec_size,
+    };
+  }
+};
+
+inline std::ostream& operator<<(std::ostream& out, const UniformInfo& info) {
+  out << "UniformInfo {" << std::endl
+      << "  uniform_name: " << info.uniform_name << std::endl
+      << "  location: " << info.location << std::endl
+      << "  type_name: " << info.type_name << std::endl
+      << "  columns: " << info.columns << std::endl
+      << "  vec_size: " << info.vec_size << std::endl
+      << "}";
+  return out;
+}
+}  // namespace
+
+TEST_P(CompilerTestRuntime, UniformsAppearInJson) {
+  ASSERT_TRUE(CanCompileAndReflect("sample_with_uniforms.frag",
+                                   SourceType::kFragmentShader,
+                                   SourceLanguage::kGLSL));
+
+  auto json_fd = GetReflectionJson("sample_with_uniforms.frag");
+  ASSERT_TRUE(json_fd);
+  nlohmann::json shader_json = nlohmann::json::parse(json_fd->GetMapping());
+  auto sampler_list = shader_json["sampled_images"];
+  auto float_list = shader_json["uniforms"];
+  ASSERT_EQ(sampler_list.size(), 2u);
+  ASSERT_EQ(float_list.size(), 6u);
+
+  {
+    // clang-format off
+    std::array expected_infos = {
+        UniformInfo::Sampler("uFirstSampler", 1u),
+        UniformInfo::Sampler("uSampler", 7u),
+    };
+    // clang-format on
+    ASSERT_EQ(sampler_list.size(), expected_infos.size());
+    for (size_t i = 0; i < expected_infos.size(); i++) {
+      EXPECT_EQ(UniformInfo::fromJson(sampler_list[i]), expected_infos[i])
+          << "index: " << i;
+    }
+  }
+
+  {
+    // clang-format off
+    std::array expected_infos = {
+        UniformInfo::Float("uFirstFloat", 0u),
+        UniformInfo::Float("uFloat", 2u),
+        UniformInfo::Vec2("uVec2", 3u),
+        UniformInfo::Vec3("uVec3", 4u),
+        UniformInfo::Vec4("uVec4", 5u),
+        UniformInfo::Mat4("uMat4", 6u),
```

Comment:
Note that the locations are not reused between samplers and float uniforms even though they are in essentially different "uniform location" spaces.